prow/cmd/deck/main.go

```diff
@@ -697,16 +700,26 @@ func initSpyglass(cfg config.Getter, o options, mux *http.ServeMux, ja *jobs.Job
 	}
 	c, err := storage.NewClient(ctx, options...)
 	if err != nil {
-		logrus.WithError(err).Fatal("Error getting GCS client")
+		// We're assuming that storage client creation must be successful when gcsCredentialsFile is set
+		// In other cases it might be okay that it fails, because the s3 client might be used instead.
+		if o.gcsCredentialsFile != "" {
+			logrus.WithError(err).Fatal("Error getting GCS client")
+		} else {
+			logrus.WithError(err).Info("Error getting GCS client, this is not a problem if only the s3 client is used")
```

Comment:
I don't agree with this. Either we want credentials and the error here is a problem, or we do not care and should be using the anonymous client.

Ignoring errors doesn't seem like a great way to handle this.

`gcsCredentialsFile == ""` + using GCS  is not an edge case, should be the best practice on GKE

Reply:
I agree with you I just have no idea how to detect this :). For example with the current code in the io package it's possible to handover neither s3-credentials nor gcs-credentials. The GCS credential discovery fails and the one from s3 works.

Ah maybe I have an idea. I guess we can make it mandatory to set `gcs-no-auth` when s3 should be used, right? So then I can throw an error whenever the gcs client creation fails?

Reply:
My case which failed was:
* s3-credentials-file was set
* gcs-credentials-file was not set

opener would have worked but process failed because the gcs client couldn't be created. So this case should be solvable by setting `gcs-no-auth`

Reply:
I'm fine with that, although it might be suboptimal for s3 users.

Another approach would be to replace `gcs-no-auth` with `gcs-require-default-creds`:
* Require creating a client with explicit credentials to succeed (aka gcsCredFile != "")
* Create a client with default credentials (aka gcsCredFile == "")
  - If this succeeds, use it
  - If it fails and `gcs-require-default-creds` then fatal
  - otherwise create an anonymous client, which should never fail

This will provide GCP users with the expected default behavior (use default creds by default) and also provide an option to blow up if default creds aren't being used.

Reply:
This would mean adding the new flag and deprecating the old one, right? I should probably also enforce that both flags (`gcs-no-auth` and `gcs-require-default-creds`) are not set at the same time.

Reply:
Because of another comment I changed the implementation to use the gcsClient inside of the opener. So now the problem doesn't exist anymore. The NewOpener implementation:
* if gcsCredentialsFile is set: throws an error if the gcp client cannot be created
* if gcsCredentialsFile is not set: sets gcsClient to nil and logs a warning

The error is then thrown the first time the gcsClient is acutally used (which won't happen for s3 paths)